src/markdown.ml

```diff
@@ -0,0 +1,341 @@
+      inline ~prefix buf md;
+      nl buf
+  | List (_attr, ty, spacing, bl) -> (
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      let tight =
+        match spacing with
+        | Loose -> false
+        | Tight -> true
+      in
+      match ty with
+      | Ordered (x, c) ->
+          add_ordered_list ?first_item_prefix ~prefix ~quote ~tight buf x c bl
+      | Bullet c ->
+          add_bullet_list ?first_item_prefix ~prefix ~quote ~tight buf c bl)
+  | Code_block (attr, label, c) ->
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      add_prefix ~first_item_prefix ~prefix buf;
+      add_code_block ~first_item_prefix ~prefix buf attr label c
+  | Thematic_break _attr ->
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      add_prefix ~first_item_prefix ~prefix buf;
+      nchar 3 buf '*';
+      nl buf
+  | Html_block (_, body) ->
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      add_prefix ~first_item_prefix ~prefix buf;
+      let n = String.length body in
+      let body =
+        if n > 0 && body.[n - 1] = '\n' then
+          String.sub body 0 (n - 1)
+        else
+          body
+      in
+      add_string_with_nl_prefix ~first_item_prefix ~prefix buf body;
+      nl buf
+  | Heading (attr, level, text) ->
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      add_prefix ~first_item_prefix ~prefix buf;
+      if 0 < level && level < 7 then (
+        nchar level buf '#';
+        sp buf
+      );
+      inline ~prefix buf text;
+      add_attrs_to_buffer ~space:true buf attr;
+      nl buf
+  | Definition_list (_attr, l) ->
+      add_nl_with_quote_prefix ~prefix ~quote ~tight buf;
+      add_prefix ~first_item_prefix ~prefix buf;
+      add_def_list ~prefix buf l
+
+and add_ordered_list ?first_item_prefix ~prefix ~quote ~tight buf x c = function
+  | [] -> ()
+  | hd :: tl ->
+      let add_list_item before_first_item_prefix i bl =
+        let n, s =
+          let s = Printf.sprintf "%d%c  " (x + i) c in
+          (String.length s, s)
+        in
+        let first_item_prefix = before_first_item_prefix ^ s in
+        let prefix = prefix ^ String.make n ' ' in
+        match bl with
+        | [] -> ()
+        | hd :: tl ->
+            block ~first_item_prefix ~prefix ~quote ~tight buf hd;
+            List.iter (fun b -> block ~prefix ~quote ~tight buf b) tl
+      in
+      add_list_item (Option.value ~default:prefix first_item_prefix) 0 hd;
+      List.iteri (fun i -> add_list_item prefix (i + 1)) tl
+
+and add_code_block ~first_item_prefix ~prefix buf attr label c =
+  let n = infer_num_backticks 3 c in
+  nchar n buf '`';
+  let label = String.trim label in
+  Buffer.add_string buf label;
+  add_attrs_to_buffer ~space:true buf attr;
+  nl buf;
+  add_prefix ~first_item_prefix ~prefix buf;
+  add_string_with_nl_prefix ~first_item_prefix ~prefix buf c;
+  let n = String.length c in
+  if n > 0 && c.[String.length c - 1] <> '\n' then begin
```

Comment:
```suggestion
  let n' = String.length c in
  if n' > 0 && c.[n'- 1] <> '\n' then begin
```

Reply:
otherwise the `nchar n buf` bellow adds too much backticks when closing the code block.